### PR TITLE
Add information on who started flow on a node.

### PR DESCRIPTION
**Pull request by kasiastreich**
Add sealed class holding information on different ways of starting a flow: RPC, peer, manual, scheduled. Needed for flow UI work.

### CODE REVIEW

**Comment by mikehearn** [comment]
Removing myself for now - Shams, can you take this please.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 1]
```diff
@@ -12,6 +12,30 @@ import org.slf4j.Logger
...
+ */
+@CordaSerializable
+sealed class CommunicationInitiator {
+    override fun toString(): String = javaClass.simpleName
+
+    abstract class WithName(open val name: String): CommunicationInitiator() {
+        override fun toString() = javaClass.simpleName + ": $name"
+    }
+
+    class Manual(override val name: String): CommunicationInitiator.WithName(name) // TODO Think of name here. What with shell access?
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: I imagine the shell will move away from starting flows directly and use the RPC. For now I think this should be called `Shell`. For tests we can use any of the existing ones - I don't like the idea of having a type just for tests which would never occur in prod code.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 2]
```diff
@@ -12,6 +12,30 @@ import org.slf4j.Logger
...
+    /** Started using [CordaRPCOps.startFlowDynamic]. Name is RPC username. */
+    class Rpc(override val name: String): CommunicationInitiator.WithName(name)
+    /** Started when we get new session initiation request. Name is peer legal name. */
+    class Peer(override val name: String): CommunicationInitiator.WithName(name)
+    class Scheduled: CommunicationInitiator()
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Make this `object`

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 3]
```diff
@@ -12,6 +12,30 @@ import org.slf4j.Logger
...
+    /** When constructing [FlowLogic] communication initiator defaults to [NotStarted]*/
+    class NotStarted: CommunicationInitiator()
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: I don't see why this should be needed - when a flow is constructed you know who created it, right?

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 4]
```diff
@@ -12,6 +12,30 @@ import org.slf4j.Logger
...
 
 /**
+ * CommunicationInitiator holds information on who started the flow. We have different ways of doing that: via RPC [CommunicationInitiator.Rpc],
+ * communication started by peer node [CommunicationInitiator.Peer], scheduled flows [CommunicationInitiator.Scheduled]
+ * or manual [CommunicationInitiator.Manual]. The last case is for all flows started in tests, shell etc. It was added
+ * added because we can start flow directly using [StateMachineManager.add] or [ServiceHubInternal.startFlow].
+ */
+@CordaSerializable
+sealed class CommunicationInitiator {
+    override fun toString(): String = javaClass.simpleName
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: This is not needed

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 5]
```diff
@@ -12,6 +12,30 @@ import org.slf4j.Logger
...
+
+    abstract class WithName(open val name: String): CommunicationInitiator() {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Nor this, as you can make these types data classes

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt (hunk 4 above)
Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: I added these methods so I display it nicely in UI. I can add a formatter for UI if you think is better.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt (hunk 5 above)
Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Same, UI.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt (hunk 4 above)
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Yeah better to have a separate formatter for the UI. `toString` is best left for debugging.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt (hunk 2 above)
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Also, the coding standard is to have a space before and after the colon when defining inheritance: `Scheduled : CommunicationInitiator`

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt (hunk 1 above)
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: It does? `IteractiveShell.runFlowByNameFragment` invokes it directly.

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Yes, it uses the "RPC interface", not the RPC wire protocol :) In future it might make sense to split the shell out into a separate worker process but for now it short-circuits the stack. However it does set the CURRENT_RPC_USER TLS slot, I think.

Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: I wasn't sure how ssh invoking will work, I will add a comment on that, as now it's done directly.
I think that more interesting case is when you can directly from flow call `serviceHub.invokeFlowAsync` which kicks off a new state machine inside flow. Is it intended? (it's also manual behaviour I am not able to classify it differently)

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt (hunk 3 above)
Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: I pass the information who started flow, when calling `startFlow, startFlowDynamic, invokeFlowAsync,StateMachineManager.add`, that's the only way I can tell if it was started through RPC/Peer/Manual. When constructing flow, I don't know that. I could make it null and then assign it.

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: The solution is quite simple I think. If you add the initiator as a property of `FlowStateMachineImpl`, that will force to you add the initiator param to `StateMachineManager.createFiber` (which is where they're created). You keep adding the param in the call chain until you reach the bits of the code where you know who the initiator is.

Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: That was my first approach, but I found carrying over that initiator param ugly, as it changes most flow starting functions on the way (only `startFlowDynamic` will remain unaffected).

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Yes it's not ideal, but it's better than exposing a `NotStarted` state. Also you're having to make the initiator field a `var`.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 6]
```diff
@@ -12,6 +12,30 @@ import org.slf4j.Logger
 import java.util.*
 
 /**
+ * CommunicationInitiator holds information on who started the flow. We have different ways of doing that: via RPC [CommunicationInitiator.Rpc],
+ * communication started by peer node [CommunicationInitiator.Peer], scheduled flows [CommunicationInitiator.Scheduled]
+ * or manual [CommunicationInitiator.Manual]. The last case is for all flows started in tests, shell etc. It was added
+ * added because we can start flow directly using [StateMachineManager.add] or [ServiceHubInternal.startFlow].
+ */
+@CordaSerializable
+sealed class CommunicationInitiator {
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Should this not be 'User'? Ultimately we do wish to check permissions and the UserService will be what we check for these.

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: That wouldn't work with flows started by peers.

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt [hunk 7]
```diff
@@ -185,7 +185,6 @@ abstract class FlowLogic<out T> {
     fun waitForLedgerCommit(hash: SecureHash): SignedTransaction {
         return stateMachine.waitForLedgerCommit(hash, this)
     }
-
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: Revert this single whitespace change

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 8]
```diff
@@ -12,6 +12,22 @@ import org.slf4j.Logger
...
+ * communication started by peer node [FlowInitiator.Peer], scheduled flows [FlowInitiator.Scheduled]
+ * or manual [FlowInitiator.Manual]. The last case is for all flows started in tests, shell etc. It was added
+ * because we can start flow directly using [StateMachineManager.add] or [ServiceHubInternal.startFlow].
+ */
+@CordaSerializable
+sealed class FlowInitiator {
+    /** Started using [CordaRPCOps.startFlowDynamic]. Name is RPC username. */
+    data class Rpc(val name: String) : FlowInitiator()
+    /** Started when we get new session initiation request. Name is peer legal name. */
+    data class Peer(val name: String) : FlowInitiator()
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Rename `name` to `legalName`

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 9]
```diff
@@ -12,6 +12,22 @@ import org.slf4j.Logger
...
 /**
+ * FlowInitiator holds information on who started the flow. We have different ways of doing that: via RPC [FlowInitiator.Rpc],
+ * communication started by peer node [FlowInitiator.Peer], scheduled flows [FlowInitiator.Scheduled]
+ * or manual [FlowInitiator.Manual]. The last case is for all flows started in tests, shell etc. It was added
+ * because we can start flow directly using [StateMachineManager.add] or [ServiceHubInternal.startFlow].
+ */
+@CordaSerializable
+sealed class FlowInitiator {
+    /** Started using [CordaRPCOps.startFlowDynamic]. Name is RPC username. */
+    data class Rpc(val name: String) : FlowInitiator()
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Would `RPC` be better?

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Rename `name` to `username`

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 10]
```diff
@@ -12,6 +12,22 @@ import org.slf4j.Logger
...
+    /** Started when we get new session initiation request. Name is peer legal name. */
+    data class Peer(val name: String) : FlowInitiator()
+    object Manual : FlowInitiator() // TODO Think of name here. What with shell access?
+    object Scheduled : FlowInitiator()
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: I think it makes sense to have the `ScheduledStateRef` as a property.

File: node/src/main/kotlin/net/corda/node/shell/InteractiveShell.kt [hunk 11]
```diff
@@ -225,8 +226,9 @@ object InteractiveShell {
         if (!FlowLogic::class.java.isAssignableFrom(clazz))
             throw IllegalStateException("Found a non-FlowLogic class in the whitelist? $clazz")
         try {
+            // TODO Now FlowInitiator.Rpc is set to ArtemisMessagingComponent.NODE_USER. Flow invocation should use startFlowDynamic.
             @Suppress("UNCHECKED_CAST")
-            val fsm = runFlowFromString({ node.services.startFlow(it) }, inputData, clazz as Class<FlowLogic<*>>)
+            val fsm = runFlowFromString({ node.services.startFlow(it, FlowInitiator.Rpc(CURRENT_RPC_USER.get().username)) }, inputData, clazz as Class<FlowLogic<*>>)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/shell/InteractiveShell.kt]: As your comment says, the shell always uses the NODE_USER for RPC so this doesn't give us as much useful information as a `Shell` initiator would.

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt [hunk 12]
```diff
@@ -471,15 +473,16 @@ class StateMachineManager(val serviceHub: ServiceHubInternal,
      *
      * Note that you must be on the [executor] thread.
      */
-    fun <T> add(logic: FlowLogic<T>): FlowStateMachine<T> {
+    @JvmOverloads
+    fun <T> add(logic: FlowLogic<T>, flowInitiator: FlowInitiator = FlowInitiator.Manual): FlowStateMachine<T> {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: There are only two uses of `add` so providing a default value for flowInitiator isn't necessary.

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 13]
```diff
@@ -92,7 +93,21 @@ interface ServiceHub : ServicesForResolution {
      *
      * @throws IllegalFlowLogicException or IllegalArgumentException if there are problems with the [logicType] or [args].
      */
-    fun <T : Any> invokeFlowAsync(logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowStateMachine<T>
+    fun <T : Any> invokeFlowAsync(logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowStateMachine<T> {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: This isn't used so it can be removed.

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 14]
```diff
@@ -92,7 +93,21 @@ interface ServiceHub : ServicesForResolution {
...
+    }
+
+    /**
+     * Will check [logicType] and [args] against a whitelist and if acceptable then construct and initiate the flow.
+     * Note that you must be on the server thread to call this method. [flowInitiator] points how flow was started,
+     * See: [FlowInitiator].
+     *
+     * @throws IllegalFlowLogicException or IllegalArgumentException if there are problems with the [logicType] or [args].
+     */
+    fun <T : Any> invokeFlowAsync(
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: You rightly pointed out that this is visible to flows. Move this to `ServiceHubInternal` to fix that.

File: node/src/main/kotlin/net/corda/node/services/api/ServiceHubInternal.kt [hunk 15]
```diff
@@ -64,14 +65,24 @@ abstract class ServiceHubInternal : PluginServiceHub {
     }
 
     /**
+     * Starts an already constructed flow. Note that you must be on the server thread to call this method. [FlowInitiator]
+     * defaults to [FlowInitiator.Manual].
+     */
+    fun <T> startFlow(logic: FlowLogic<T>): FlowStateMachine<T> = startFlow(logic, FlowInitiator.Manual)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/api/ServiceHubInternal.kt]: This is only used in tests so I suggest you mark it with `@VisibleForTesting`. Also use a hard coded Rpc user like `Only For Testing`. This will then remove the need for Manual completely. Ideally this method should be an extension method in the test utils class, but this will do for now.

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt (hunk 13 above)
Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: Is there a better way of overloading functions with default parameters? As `@JvmOverloads` doesn't work in that case - you cannot add annotation on interface, on implementation you are not able specify default parameter.

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: But why is a default parameter even needed?

Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: I assumed that it's part of public API so it would be nice, but as you suggested I will move it to `ServiceHubInternal`.

File: core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt [hunk 16]
```diff
@@ -185,7 +185,7 @@ abstract class FlowLogic<out T> {
...
+
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowLogic.kt]: You've added whitespace

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 17]
```diff
@@ -12,6 +13,23 @@ import org.slf4j.Logger
...
 
 /**
+ * FlowInitiator holds information on who started the flow. We have different ways of doing that: via RPC [FlowInitiator.RPC],
+ * communication started by peer node [FlowInitiator.Peer], scheduled flows [FlowInitiator.Scheduled]
+ * or manual [FlowInitiator.Manual]. The last case is for all flows started in tests, shell etc. It was added
+ * because we can start flow directly using [StateMachineManager.add] or [ServiceHubInternal.startFlow].
+ */
+@CordaSerializable
+sealed class FlowInitiator {
+    /** Started using [CordaRPCOps.startFlowDynamic]. Name is RPC username. */
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: The second part of this comment is not relevant. Likewise below.

File: node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt [hunk 18]
```diff
@@ -359,7 +361,7 @@ class StateMachineManager(val serviceHub: ServiceHubInternal,
 
         val session = try {
             val flow = flowFactory(sender)
-            val fiber = createFiber(flow)
+            val fiber = createFiber(flow, FlowInitiator.Peer(sender.name))
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt]: I wonder if it's better to have the `Party` instead of just the name

File: client/rpc/src/integration-test/kotlin/net/corda/client/rpc/CordaRPCClientTest.kt [hunk 19]
```diff
@@ -115,4 +111,38 @@ class CordaRPCClientTest : NodeBasedTest() {
...
+    fun `flow initiator via RPC`() {
+        val proxy = createRpcProxy(rpcUser.username, rpcUser.password)
+        val smUpdates = proxy.stateMachinesAndUpdates()
+        var countRpcFlows = 0
+        var countShellFlows = 0
+        smUpdates.second.subscribe {
+            if (it is StateMachineUpdate.Added) {
+                val initiator = it.stateMachineInfo.initiator
+                if (initiator is FlowInitiator.RPC)
+                    countRpcFlows += 1
```
Comment (shamsasari) [inline client/rpc/src/integration-test/kotlin/net/corda/client/rpc/CordaRPCClientTest.kt]: `++`

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 20]
```diff
@@ -12,6 +13,23 @@ import org.slf4j.Logger
...
+    data class RPC(val username: String) : FlowInitiator()
+    /** Started when we get new session initiation request. Name is peer legal name. */
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Perhaps we should start fixing up our User concept. Perhaps make a UserIdentity data class for this username (with a string content for now) and do same on the username property of User class. Long term I am sure that will end up a user key, or token.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt [hunk 21]
```diff
@@ -12,6 +13,23 @@ import org.slf4j.Logger
...
+    data class Peer(val legalName: String) : FlowInitiator()
+    /** Started as scheduled activity. */
```
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: This should be wrapped around Party as Shams suggested and to prevent clashes with the Identity work.

File: core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt (hunk 20 above)
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: Changing `User` will ripple changes everywhere. `User` was a simple solution to get RPC security working and was never meant to be the proper solution. This deserves it's own PR at some point.

Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/flows/FlowStateMachine.kt]: I already did the change Matthew requested. I could split it and raise as new PR on Monday.

**Comment by kasiastreich** [comment]
@mikehearn could you comment on `invokeFlowAsync` (from my previous comment: you can directly call from flow `serviceHub.invokeFlowAsync` which kicks off a new state machine inside flow. Is it intended?). I saw some discussion on public slack regarding this. Currently I moved that method to `ServiceHubInternal`.

**Comment by mikehearn** [comment]
Hmm. It's neither intended nor unintended right now. It's an API design point we need to consider. For now the move is fine. If we decide to allow the use case of one flow starting a new top level flow, we should at least write test cases for it and document it.
